# Assignment 1 - Optimization Methods
## Matteo Ghilardini
---

### Part 1: exercise

1. __What is the definition of a symmetric matrix?__<br>
> A matrix $M \in R^{n \times n}$ is symmetric if is squared, and any element of $M$ satisfy $M_{ij} == M_{ji}$ (i.e. row $x$ is equal to column $x$ $\forall x \ge 1 \land x \le n$)

2. __What are the definitions of eigenvalues and eigenvectors?__ <br>
> A scalar $\lambda$ is called **eigenvalue** of a matrix $M \in R^{n \times n}$ if there exist a non-zero vector $v$ (called **eigenvector**) such that holds $Mx=\lambda x$.  
> In other words, a non-zero vector $v$ is called **eigenvector** if by multiplying it to a matrix $M$, it results to a scaled version of $v$ (the scaling factor is the **eigenvalue** $\lambda$).

3. __What is a singular matrix?__ <br>
> A matrix $M \in R^{n \times n}$ is singular if is not invertible, i.e. the determinant of $M$ is zero.

4. __What is the definition of a positive definite matrix? definition of a positive semi- definite matrix?__ <br>
> A matrix $M \in R^{n \times n}$ is **positive defined** if it is symmetric and for all nonzero vectors $x \in \mathbb{R}^n$: $x^T A x > 0$; implying that all **eigenvalues** of $M$ are strictly positive.  <br>
> Otherwise, a matrix $M \in R^{n \times n}$ is **semi-positive defined** if it is symmetric and and for all nonzero vectors $x \in \mathbb{R}^n$: $x^T A x \ge 0$; implying that  all **eigenvalues** of $M$ are non-negative.

5. __Let $M \in R^{n×n}$ a nonsingular square matrix and let $A = M^T M$ (here the supscript $^T$ means that we take the transpose). Prove that $A$ is positive definite.__ _(Hint: recall that $x^Tx=||x||^2 \;for \; x \in R^n$)_
> In order to proove that $A$ is positive defined we have to proove that $A$ is symmetric and for all nonzero vectors $x \in \mathbb{R}^n$: $x^T A x > 0$  
> Since $A = M^T M$, we can write $x^T A x = x^T (M^T M) x > 0$  
> By associativity become $x^T (M^T M) x = (M x)^T (M x)$  
> Since $Mx$ is a vector, the doct product between a vector and itself (or in this case, itself transposed) is equivalent to the squared norm: $|Mx|^2$  
> Since $M x$ is nonzero for any nonzero $x$, the squared norm $\| M x \|^2 > 0$ for all $x \neq 0$. Thus, $x^T A x > 0$, **proving that $A$ is positive definite**.

### Part 2: programming problem

#### Setup the environment:

Install numpy in the jupiter notebook environment:

In [1]:
pip install numpy

Note: you may need to restart the kernel to use updated packages.


Import _numpy_ library

In [2]:
import numpy as np

#### Programming solutions

##### 1. Create a vector b with the elements
$ b=
    \begin{bmatrix}
        -3 \\
        -8 \\
        7
    \end{bmatrix}
$
and print it.

In [3]:
b = np.array([[-3, -8, 7]]) # 2D vector array, i.e. matrix with 1 row and 3 columns
b = b.T # Convert the 2D array from a matrix 1x3 to a matrix 3x1, i.e. a column vector
print(b)
# Otherwise the same result can also be obtained by using the following code, but 
#  is simply a vector without consider if is row-vector or column-vector:
# b = np.array([-3, -8, 7])
# print(b)

[[-3]
 [-8]
 [ 7]]


##### 2. Create a matrix A with the elements 
$ A=
    \begin{bmatrix}
        -10 & 3 & 11 \\
        3 & -5 & -4 \\
        11 & -4 & -7
    \end{bmatrix}
$
and print it.

In [4]:
A = np.array([
    [-10, 3, 11], 
    [3, -5, -4], 
    [11, -4, -7]
]) # 2D array, i.e. matrix with 3 rows and 3 columns
print(A)

[[-10   3  11]
 [  3  -5  -4]
 [ 11  -4  -7]]


##### 3. Is the matrix A symmetric? Write your answer in a cell in the notebook and then check it with Python.

*Yes, a matrix $M$ is symmetric when $M = M^T$, or in other words, any element of $M$ $M_{ij}$ is equal to the element $M_{ji}$. Here is the check:*

In [5]:
def is_symmetric(matrix):
    return np.array_equal(matrix, matrix.T)

print(is_symmetric(A))

True


##### 4. Compute $x$ as the solution of the linear system $Ax = b$ and print it.

In [6]:
x = np.linalg.solve(A,b)
print(x)

[[2.]
 [2.]
 [1.]]


##### 5. To check if the solution is correct, compute $r$ as the product $Ax$ and print it.

In [7]:
# I first tryed using r=A*x, but then i found out that this performs a element-wise 
#   multiplication, not a matrix multiplication. So I've searched on numpy documentation, 
#   and found out that np.dot() is the function to use for matrix multiplication.
r = np.dot(A, x) 
print(r)

[[-3.]
 [-8.]
 [ 7.]]


##### 6. Print the difference of $r$ and $b$ as $r−b$. (This should be zero. Is it? Briefly discuss this in the cell for this question.)

In [8]:
print(np.linalg.norm(r-b)) # Normalizing the difference between the two vectors we get ||r-b||

3.580361673049448e-15


*We find out that the result obtained is very close to 0 ($3.51 * 10^{-15}$) but is not exactly 0 probably due to floating-point approximations done in computations*

##### 7. Print the determinant of A.

In [9]:
print(np.linalg.det(A))

214.0000000000001


##### 8. Repeat questions 2-7 for the matrix A1 with the elements
$ A1=
    \begin{bmatrix}
        -1 & 3000 & 1 \\
        2000 & -1.5 & -2.5 \\
        -1999 & −2998.5 & 1.5
    \end{bmatrix}
$


In a cell, describe and explain what you observe.

2. Create a matrix A1 with the elements 
$ A1=
    \begin{bmatrix}
        -1 & 3000 & 1 \\
        2000 & -1.5 & -2.5 \\
        -1999 & -2998.5 & 1.5
    \end{bmatrix}
$
and print it.

In [10]:
A1 = np.array([
    [-1, 3000, 1],
    [2000, -1.5, -2.5],
    [-1999, -2998.5, 1.5]
])
print(A1)

[[-1.0000e+00  3.0000e+03  1.0000e+00]
 [ 2.0000e+03 -1.5000e+00 -2.5000e+00]
 [-1.9990e+03 -2.9985e+03  1.5000e+00]]


3. Is the matrix A symmetric? Write your answer in a cell in the notebook and then check it with Python.

*No, the matrix is symmetric since $A1 \not= A1^T$. Here is the check:*

In [11]:
print(is_symmetric(A1))

False


4. Compute $x$ as the solution of the linear system $Ax = b$ and print it.

In [12]:
x1 = np.linalg.solve(A1,b)
print(x1)

[[ 2.25135002e+13]
 [-5.99729500e+12]
 [ 1.80143985e+16]]


5. To check if the solution is correct, compute $r$ as the product $Ax$ and print it.

In [13]:
r1 = np.dot(A1, x1) 
print(r1)

[[-4.]
 [-8.]
 [ 8.]]


6. Print the difference of $r$ and $b$ as $r−b$. (This should be zero. Is it? Briefly discuss this in the cell for this question.)

In [14]:
print(np.linalg.norm(r1-b)) # Normalizing the difference between the two vectors we get ||r-b||

1.4142135623730951


*We find out that the result obtained is much more bigger than the one obtained with the matrix $M$. This highlight the fact that is very likely to have an error in computations, due to the fact that the matrix is **Ill-Conditioned**, i.e. $det(M1) \approx 0$*

7. Print the determinant of A

In [15]:
print(np.linalg.det(A1))

1.3322672964832825e-09


*Looking at the determinant of $A1$ we notice that is very close to zero ($1.33 * 10^{-9}$) confirming that the matrix $A1$ is **Ill-Conditioned** and so susceptible to errors in computations (as hypothesized in task 8.6).*